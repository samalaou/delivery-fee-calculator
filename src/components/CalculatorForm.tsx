import { Button, Grid } from '@mui/material';
import { DateTimePicker } from '@mui/x-date-pickers';
import dayjs, { Dayjs } from 'dayjs';
import { ChangeEvent, FormEvent, useState } from 'react';

import NumberField from './NumberField';

interface CalculatorFormProps {
  calculateDeliveryFee(
    cartValue: number,
    distance: number,
    numberOfItems: number,
    orderTime: Dayjs,
  ): void;
}

interface FormValues {
  cartValue: number | '';
  distance: number | '';
  numberOfItems: number | '';
  orderTime: Dayjs;
}

interface FormErrors {
  cartValue: boolean;
  distance: boolean;
  numberOfItems: boolean;
}

function CalculatorForm({
  calculateDeliveryFee,
}: CalculatorFormProps): JSX.Element {
  const [formValues, setFormValues] = useState<FormValues>({
    cartValue: '',
    distance: '',
    numberOfItems: '',
    orderTime: dayjs(new Date()),
  });

  const [errors, setErrors] = useState<FormErrors>({
    cartValue: false,
    distance: false,
    numberOfItems: false,
  });

  const handleChange = (e: ChangeEvent<HTMLInputElement>): void => {
    const { name, value } = e.target;
    const numericValue = parseFloat(value);
    setFormValues((prevValues) => ({ ...prevValues, [name]: numericValue }));
    setErrors((prevErrors) => ({ ...prevErrors, [name]: isNaN(numericValue) }));
  };

  const handleOrderTimeChange = (date: Dayjs | null) => {
    if (date !== null) {
      setFormValues((prevValues) => ({ ...prevValues, orderTime: date }));
    }
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    const parsedOrderTime = formValues.orderTime
      ? formValues.orderTime
      : dayjs(new Date());
    calculateDeliveryFee(
      parseFloat(formValues.cartValue as string) || 0,
      parseFloat(formValues.distance as string) || 0,
      parseFloat(formValues.numberOfItems as string) || 0,
      parsedOrderTime,
    );
  };

  return (
    <form onSubmit={handleSubmit} data-test-id="calculatorForm">
      <NumberField
        value={formValues.cartValue}
        handleChange={handleChange}
        id="cartValue"
        endAdornment="â‚¬"
        dataTestId="cartValue"
        name="cartValue"
        label="Cart value"
        error={errors.cartValue}
      />
      <NumberField
        value={formValues.distance}
        handleChange={handleChange}
        id="distance"
        endAdornment="m"
        dataTestId="deliveryDistance"
        name="distance"
        label="Delivery distance"
        error={errors.distance}
      />
      <NumberField
        value={formValues.numberOfItems}
        handleChange={handleChange}
        id="numberOfItems"
        dataTestId="numberOfItems"
        name="numberOfItems"
        label="Number of items"
        decimalScale={0}
        error={errors.numberOfItems}
      />
      <Grid item xs={12} style={{ margin: '20px 10px' }}>
        <DateTimePicker
          label="Time of order"
          name="orderTime"
          value={formValues.orderTime}
          data-test-id="orderTime"
          onChange={handleOrderTimeChange}
          slotProps={{
            textField: {
              fullWidth: true,
              inputProps: { 'data-test-id': 'orderTime' },
            },
          }}
          sx={{ backgroundColor: 'white' }}
        />
      </Grid>
      <Grid item xs={12} style={{ margin: '20px 10px' }}>
        <Button
          fullWidth
          color="secondary"
          size="large"
          type="submit"
          variant="contained"
        >
          Calculate
        </Button>
      </Grid>
    </form>
  );
}

export default CalculatorForm;
