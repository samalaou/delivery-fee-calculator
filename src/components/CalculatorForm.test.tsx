import { render, fireEvent } from '@testing-library/react';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';

import CalculatorForm from './CalculatorForm';
import { LocalizationProvider } from '@mui/x-date-pickers';
import dayjs from 'dayjs';

describe('CalculatorForm', () => {
  const mockCalculateDeliveryFee = jest.fn();

  beforeEach(() => {
    mockCalculateDeliveryFee.mockClear();
  });
  it('matchs snapshot', () => {
    const { container, getByTestId } = render(
      <LocalizationProvider dateAdapter={AdapterDayjs}>
        <CalculatorForm calculateDeliveryFee={mockCalculateDeliveryFee} />
      </LocalizationProvider>,
    );
    const newOrderTime = '01/26/2024 05:16 PM';
    fireEvent.change(getByTestId('orderTime'), {
      target: { value: newOrderTime },
    });
    expect(container).toMatchSnapshot();
  });

  it('submits the form with correct values when Calculate button is clicked', () => {
    const { getByTestId } = render(
      <LocalizationProvider dateAdapter={AdapterDayjs}>
        <CalculatorForm calculateDeliveryFee={mockCalculateDeliveryFee} />
      </LocalizationProvider>,
    );

    // Fill the form
    fireEvent.change(getByTestId('cartValue'), { target: { value: '50' } });
    fireEvent.change(getByTestId('deliveryDistance'), {
      target: { value: '10' },
    });
    fireEvent.change(getByTestId('numberOfItems'), { target: { value: '3' } });

    // Submit the form
    fireEvent.submit(getByTestId('calculatorForm'));

    // Check if calculateDeliveryFee was called with the correct arguments
    expect(mockCalculateDeliveryFee).toHaveBeenCalledWith(
      50,
      10,
      3,
      expect.any(dayjs),
    );
  });

  it('updates form values when inputs are changed', () => {
    const { getByTestId } = render(
      <LocalizationProvider dateAdapter={AdapterDayjs}>
        <CalculatorForm calculateDeliveryFee={mockCalculateDeliveryFee} />
      </LocalizationProvider>,
    );

    // Fill the form
    fireEvent.change(getByTestId('cartValue'), { target: { value: '20' } });
    fireEvent.change(getByTestId('deliveryDistance'), {
      target: { value: '1500' },
    });
    fireEvent.change(getByTestId('numberOfItems'), { target: { value: '5' } });
    const newOrderTime = '01/26/2024 05:16 PM';
    fireEvent.change(getByTestId('orderTime'), {
      target: { value: newOrderTime },
    });

    // Check if form values are updated
    expect(getByTestId('cartValue')).toHaveValue('20');
    expect(getByTestId('deliveryDistance')).toHaveValue('1500');
    expect(getByTestId('numberOfItems')).toHaveValue('5');
    expect(getByTestId('orderTime')).toHaveValue(newOrderTime);
  });
});
