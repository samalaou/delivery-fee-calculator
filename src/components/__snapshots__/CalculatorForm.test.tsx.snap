// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorForm matchs snapshot 1`] = `
<div>
  <form
    data-test-id="calculatorForm"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
      style="margin: 20px 10px;"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        inputmode="numeric"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="cartValue"
          id="cartValue-label"
        >
          Cart value
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            data-test-id="cartValue"
            id="cartValue"
            min="0"
            name="cartValue"
            notched="true"
            required=""
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
             
            €
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Cart value
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
      style="margin: 20px 10px;"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        inputmode="numeric"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="distance"
          id="distance-label"
        >
          Delivery distance
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            data-test-id="deliveryDistance"
            id="distance"
            min="0"
            name="distance"
            notched="true"
            required=""
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
             
            m
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Delivery distance
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
      style="margin: 20px 10px;"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        inputmode="numeric"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="numberOfItems"
          id="numberOfItems-label"
        >
          Number of items
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            data-test-id="numberOfItems"
            id="numberOfItems"
            min="0"
            name="numberOfItems"
            notched="true"
            required=""
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
             
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Number of items
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
      style="margin: 20px 10px;"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wxbnyc-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r3:"
          id=":r3:-label"
        >
          Time of order
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            data-test-id="orderTime"
            id=":r3:"
            inputmode="text"
            name="orderTime"
            placeholder="MM/DD/YYYY hh:mm aa"
            type="text"
            value="01/26/2024 05:16 PM"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 26, 2024"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Time of order
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
      style="margin: 20px 10px;"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-rggcag-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Calculate
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
